[85zrq2dfh] Add API's for Product/Product Version CRUD

BattleBaxx opened the pull request:
- Adds APIs for Product/Product Version CRUD

BattleBaxx committed:
Add: is_deleted column for product table

BattleBaxx committed:
Add: doa for products route
Add: Bad request Exception
Add: Schemas for Product

BattleBaxx committed:
Add: Routes and Services for product API

RazCrimson commented on kori/__main__.py:
```diff
@@ -2,6 +2,7 @@
 from fastapi import FastAPI
 
 from kori.app.core.config import Settings
+from kori.app.routers.product import product_router
```
I see that the linter has not been run.

Please fix it

EDIT: Its broken in many other locations, make sure to fix all files

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,161 @@
+    )
+
+    try:
+        session.add(new_product_db)
```
Can we have both the inserts (product & productVersion) in a transaction?

Don't want a case where a product got inserted and the version insert failed for some reason and we get a weird state for the product.

RazCrimson replied:
Would prefer a Demarcated Transaction rather than an implicit one.

BattleBaxx replied:
Done

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,161 @@
+    if product.is_deleted:
+        raise NotFoundException(message="Product is deleted")
+
+    product_version = list(session.query(ProductVersion).filter(ProductVersion.product_id == product_id).order_by(
```
Rather than listing, can you just try to fetch the record that is equal to 9999-12-31..... or add filter to look for records with a `validTo` greater than the current timestamp

RazCrimson replied:
You can use `query.first()` instead of `list(query)[0]` in many cases

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,161 @@
+    session = db_connector.get_session()
+    product = list(session.query(Product).filter(Product.id == product_id))
+
+    if product and product[0].is_deleted: raise NotFoundException(message="Product Details not found")
```
if a product is not found, what will happen? 🙁

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,161 @@
+def update(product_id: UUID, product_data: ProductUpdate) -> ProductSchema:
+    session = db_connector.get_session()
+
+    existing_product = list(session.query(Product).filter(Product.id == product_id))[0]
```
What will happen when an update request happens for a product that does not exist?

Ans: An `TypeError` with a nice code 500

RazCrimson commented on kori/alembic/versions/c48eb23a40bc_create_product_table.py:
```diff
@@ -24,6 +24,7 @@ def upgrade() -> None:
         sa.Column("id", UUID(as_uuid=True), primary_key=True, default=uuid.uuid4),
         sa.Column("org_id", UUID(as_uuid=True), nullable=False),
         sa.Column("reorder_level", sa.Integer(), nullable=False),
+        sa.Column("is_deleted", sa.Boolean(), default=False),
```
Just to confirm, were these changes propagated to the database?

BattleBaxx replied:
If you mean in the models, then yes

RazCrimson replied:
I mean the actual database.
As in did u go update the live database in azure with this constraint?
Also we are missing the migrations

BattleBaxx replied:
Yes, tested too

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,161 @@
+
+    existing_product = list(session.query(Product).filter(Product.id == product_id))[0]
+
+    latest_product_version = \
```
What happens when there are no version?
That case should not occur logically but having a explicit error would be helpful when debugging.

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,161 @@
+    }
+    version_id = latest_product_version.version_id
+
+    if not (updated_product_version_data["name"] == latest_product_version.name and updated_product_version_data[
```
Imo, can be rewritten in a simpler manner with a comprehension.

RazCrimson commented on kori/app/routers/product.py:
```diff
@@ -0,0 +1,35 @@
+def get_product_by_id(product_id: UUID, timestamp: Optional[datetime] = None) -> ProductSchema:
+    if timestamp:
+        return product_service.get_product_by_timestamp(product_id=product_id, timestamp=timestamp)
+    else:
```
Redundant else

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,161 @@
+    )
+
+
+def get_product_by_timestamp(product_id: UUID, timestamp: datetime) -> ProductSchema:
```
Why not have one method to fetch the product and if the timestamp is `None` we use `datetime.now()` instead?

Same for the API too

RazCrimson commented on kori/app/schemas/product.py:
```diff
@@ -0,0 +1,28 @@
+    org_id: UUID
+
+
+class ProductSchema(ProductCreate):
```
Would recommend adding a timestamp, as productSchema is snapshot of ProductSchema at a particular point in time.

BattleBaxx replied:
Should it just be current time?

RazCrimson replied:
For latest, yea use current time. 
If they request for some timestamp, then use that timestamp

RazCrimson commented on kori/app/services/product.py:
```diff
@@ -0,0 +1,26 @@
+from kori.app.schemas.product import ProductCreateRequest, ProductSchema, ProductCreate, ProductUpdate
+
+
+def create(org_id: UUID, product_data: ProductCreateRequest) -> ProductSchema:
```
If the service layer is going to be this simple, no harm is skipping it and just directly interact with DAO until the code has enough logic for a service layer.

No need to change it now, just mentioning as a bit of info.

BattleBaxx replied:
Noted

BattleBaxx committed:
Refactor: Run pre-commit

BattleBaxx committed:
Refactor: use .one() instead of list
Refactor: Use largest time to query latest product version
Add: Exception handling for all queries
Add: Product and its version in same transaction

BattleBaxx committed:
Refactor: use single get function instead of two

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,152 @@
+        if timestamp:
+            product_version = (
+                session.query(ProductVersion)
+                .filter(ProductVersion.valid_from <= timestamp)
```
Use a single filter clause
Ref: https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.filter

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,152 @@
+        else:
+            product_version = (
+                session.query(ProductVersion)
+                .filter(ProductVersion.product_id == product_id)
```
Use a single filter clause

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,152 @@
+
+        latest_product_version = (
+            session.query(ProductVersion)
+            .filter(ProductVersion.product_id == product_id)
```
Use a single filter clause

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,152 @@
+        if product.is_deleted:
+            raise NotFoundException(message="Product is deleted")
+
+        if timestamp:
```
Default the timestamp to the current time, if timestamp is None.
Ex: 
```py
if timestamp is None:
    timstamp = datetime.now()
```

Now we just need to write the query once, thereby reducing the conditional and also lesser lines of code 😏

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,152 @@
+def update(product_id: UUID, product_data: ProductUpdate) -> ProductSchema:
+    session = db_connector.get_session()
+    try:
+        existing_product = session.query(Product).filter(Product.id == product_id).one()
```
Note: You can just do `session.query(Product).get(product_id)` instead of `one` when you are filtering by primary keys. Make sure to follow it with a null check

Ref: https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.get

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,152 @@
+def get_product(product_id: UUID, timestamp: Optional[datetime]) -> ProductSchema:
+    session = db_connector.get_session()
+    try:
+        product = session.query(Product).filter(Product.id == product_id).one()
```
Note: You can just do session.query(Product).get(product_id) instead of one when you are filtering by primary keys. Make sure to follow it with a null check

Ref: https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.get

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,152 @@
+    }
+    version_id = latest_product_version.version_id
+
+    if not (
```
Remember boolean algebra?

Can be re-written as
```py
if (
        updated_product_version_data["name"] != latest_product_version.name
        or updated_product_version_data["price"] != latest_product_version.price
        or updated_product_version_data["measurement_unit"] != latest_product_version.measurement_unit
    ):
```

Is more efficient as it matches fast rather than checking each condition.

Imo, the best solution for this would be a generator based evaluation with `any`
Assuming both are dicts
`if any(a[k] != b[k] for k in ["name", "measurement", "price"]):`

RazCrimson commented on tests/manual/db/sqlalchemy_test.py:
```diff
@@ -120,17 +120,17 @@
-    session.delete(store)
-    session.delete(gc)
-    session.delete(org)
+    # session.delete(product_billed)
```
??????????

BattleBaxx committed:
Refactor: use demarcated transation

BattleBaxx committed:
Add: timestamp in ProductSchema

BattleBaxx committed:
Refactor: use a single filter clause

BattleBaxx committed:
Refactor: make timestamp default current time

BattleBaxx committed:
refactor: restructure query to get product using primary key

BattleBaxx committed:
fix: uncomment delete

BattleBaxx committed:
refactor: comparision for prev product version

BattleBaxx committed:
add: null check after get product query

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,157 @@
+
+    with session.begin() as transaction:
+        try:
+            session.add(new_product_db)
```
Shouldn't it be `transaction.add` instead of  `session.add`

RazCrimson commented on kori/app/dao/product.py:
```diff
@@ -0,0 +1,157 @@
+    )
+
+
+def get_product(product_id: UUID, timestamp: Optional[datetime] = datetime.now()) -> ProductSchema:
```
The default value should not be specified like this. 
This will cause the default value to be same time as when the application was started.

Make the default value as `None` and have a if condition inside that checks for None and assigns it the value `datetime.now()`

BattleBaxx committed:
fix: change default values for timestamp to None
